File: src/rockstor/storageadmin/views/snapper.py
"""
Views for all things related to snapper
"""

from dbus import DBusException
from rest_framework.response import Response
from rest_framework.exceptions import NotFound
from rest_framework import status
from system.snapper import Snapper
import rest_framework_custom as rfc

"""
Copyright (c) 2016 RockStor, Inc. <http://rockstor.com>
This file is part of RockStor.

RockStor is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published
by the Free Software Foundation; either version 2 of the License,
or (at your option) any later version.

RockStor is distributed in the hope that it will be useful, but
WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program. If not, see <http://www.gnu.org/licenses/>.
"""

# Obtain snapper interface
snapper = Snapper()


class SnapperConfigList(rfc.GenericAPIView):
    """List all snapper configurations or create a new one.
    """
    def get(self, request):
        return Response(snapper.config_list())

    def post(self, request):
        with self._handle_exception('Failed to create configuration'):
            name = request.data.get('NAME')
            subvolume = request.data.get('SUBVOLUME')
            snapper.CreateConfig(name, subvolume, 'btrfs', 'default')
        return Response(request.data, status=status.HTTP_201_CREATED)


class SnapperConfigDetail(rfc.GenericAPIView):
    """Create/edit/delete a snapper configuration.
    """
    def get(self, request, name):
        try:
            config = snapper.get_config(name)
        except DBusException:
            raise NotFound('Configuration \'%s\' not found.' % name)
        else:
            return Response(config)

    def put(self, request, name):
        try:
            config = snapper.get_config(name)
        except DBusException:
            # Config does not exist
            with self._handle_exception(request):
                subvolume = request.data.get('SUBVOLUME')
                snapper.CreateConfig(name, subvolume, 'btrfs', 'default')

        # Ignore the read-only properties
        config = {key: value for key, value in request.data.items()
                  if key not in ['NAME', 'SUBVOLUME', 'FSTYPE']}
        snapper.SetConfig(name, config)
        return Response(request.data)

    def delete(self, request, name):
        with self._handle_exception('Failed to delete configuration'):
            snapper.DeleteConfig(name)
        return Response(status=status.HTTP_204_NO_CONTENT)


def extract_snapshot_data(request):
    """Extract the right data for creating/updating snapshots from a request.
    """
    return [request.data.get(item) for item in
            ('description', 'cleanup', 'userdata')]


class SnapperSnapshotList(rfc.GenericAPIView):
    """Return the list of snapshots for the given snapper configuration, or
    create a new snapshot.
    """
    def get(self, request, name):
        try:
            snapshot_list = snapper.list_snapshots(name)
        except DBusException:
            raise NotFound('Configuration \'%s\' not found.' % name)
        else:
            return Response(snapshot_list)

    def post(self, request, name):
        number = 0
        with self._handle_exception(request):
            number = snapper.CreateSingleSnapshot(
                name, *extract_snapshot_data(request)
            )
        return Response(snapper.get_snapshot(name, number),
                        status=status.HTTP_201_CREATED)


class SnapperSnapshotDetail(rfc.GenericAPIView):
    """Retrieve, update or delete a particular snapshot.
    """
    def get(self, request, name, number):
        try:
            snapshot = snapper.get_snapshot(name, number)
        except DBusException as e:
            raise NotFound(e)
        else:
            return Response(snapshot)

    def put(self, request, name, number):
        with self._handle_exception(request):
            snapper.SetSnapshot(name, number, *extract_snapshot_data(request))
        return Response(request.data, status=status.HTTP_201_CREATED)

    def delete(self, request, name, number):
        try:
            snapper.DeleteSnapshots(name, [number])
            snapper.Sync(name)
        except DBusException as e:
            raise NotFound(e)
        else:
            return Response(status=status.HTTP_204_NO_CONTENT)

File: src/rockstor/storageadmin/static/storageadmin/js/views/snapper.js
/*
 *
 * @licstart  The following is the entire license notice for the
 * JavaScript code in this page.
 *
 * Copyright (c) 2016 RockStor, Inc. <http://rockstor.com>
 * This file is part of RockStor.
 *
 * RockStor is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published
 * by the Free Software Foundation; either version 2 of the License,
 * or (at your option) any later version.
 *
 * RockStor is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 *
 * @licend  The above is the entire license notice
 * for the JavaScript code in this page.
 *
 */

var SnapperMainView = RockstorLayoutView.extend({
    events: {
        'click button#config-delete': 'onConfigDeleteClicked',
        'click button#select': function() {
            this.table.rows().select();
        },
        'click button#unselect': function() {
            this.table.rows().deselect();
        },
        'click button#delete': 'onDeleteClicked'
    },

    initialize: function() {
        this.template = window.JST.snapper_main;
        this.collection = new SnapperConfigCollection();
        this.selector = new BootstrapSelect({
            collection: this.collection
        });
        this.selectedSnapshots = [];
        this.listenTo(this.collection, 'change:snapshots', this.updateSnapshots);
        this.listenTo(this.selector, 'change:selection', this.updateSelectedConfig);
        this.collection.fetch();
    },

    render: function() {
        this.$el.html(this.template);
        this.selector.setElement(this.$('select#config'));
        this.table = this.$('table').DataTable({
            order: [[2,'asc']],
            select: true,
            columns: [
                {data: 'number', title: 'ID'},
                {data: 'type', title: 'Type'},
                {data: 'timestamp', title: 'Start Time'},
                {data: 'end_time', title: 'End Time'},
                {data: 'description', title: 'Description'},
                {data: 'cleanup', title: 'Cleanup Algorithm'},
                {data: 'userdata', title: 'User Data'}
            ]
        });

        // Enable snapshot delete button only when there is a selection
        var deleteButton = this.$('#delete');
        disableButton(deleteButton);
        this.table.on('select.dt deselect.dt', function(event, dt, type, indexes) {
            if (dt.rows({selected: true}).count())
                enableButton(deleteButton);
            else
                disableButton(deleteButton);
        });
        return this;
    },

    updateSelectedConfig: function(config) {
        this.$('h1 > span').text('for ' + config.get('SUBVOLUME'));
        config.snapshots.fetch();
    },

    updateSnapshots: function(config) {
        var data = [];
        config.snapshots.each(function(snapshot) {
            // Do some processing to combine pre/post snapshots into a single row
            var current = _.clone(snapshot.attributes);
            current.end_time = '';
            current.userdata = _.map(current.userdata, function(value, key) {
                return key + '=' + value;
            }).join(', ');
            if (current.pre_number) {
                var pre = data[data.length-1];
                pre.number = pre.number + ', ' + current.number;
                pre.type += ', post';
                pre.end_time = current.timestamp;
            } else {
                data.push(current);
            }
        });
        this.table.clear().rows.add(data).draw();
    },

    onDeleteClicked: function() {
        // Filter out the list of IDs from the table selection.
        this.selectedSnapshots = this.table
            .rows({selected: true})
            .data()
            .pluck('number')
            .join(' ')
            .match(/\d+/g);

        // Show confirmation dialog if multiple snapshots are to be deleted at
        // once.
        if (!this.selectedSnapshots) {
            return;
        } else if (this.selectedSnapshots.length == 1) {
            this.deleteSnapshots();
        } else {
            var dialog = new WarningDialog({
                el: this.$('.snapper-modal'),
                message: 'You are about to delete multiple snapshots. ' +
                    'This action cannot be undone. Proceed?'
            }).render();
            dialog.once('submit', this.deleteSnapshots, this);
        }
    },

    deleteSnapshots: function() {
        this.selector.model.snapshots.remove(this.selectedSnapshots);
    },

    onConfigDeleteClicked: function() {
        var dialog = new WarningDialog({
            el: this.$('.snapper-modal'),
            message: 'All snapshots associated with the selected configuration ' +
                'will also be deleted. This action cannot be undone. Proceed?'
        }).render();
        dialog.once('submit', function() {
            this.selector.model.destroy();
        }, this);
    }
});

